api/host.py

```diff
@@ -148,105 +122,26 @@ def get_host_list(
     order_how=None,
     staleness=None,
 ):
-    if fqdn:
-        query = find_hosts_by_canonical_facts(current_identity.account_number, {"fqdn": fqdn})
-    elif display_name:
-        query = find_hosts_by_display_name(current_identity.account_number, display_name)
-    elif hostname_or_id:
-        query = find_hosts_by_hostname_or_id(current_identity.account_number, hostname_or_id)
-    elif insights_id:
-        query = find_hosts_by_canonical_facts(current_identity.account_number, {"insights_id": insights_id})
-    else:
-        query = Host.query.filter(Host.account == current_identity.account_number)
-
-    if tags:
-        # add tag filtering to the query
-        query = find_hosts_by_tag(current_identity.account_number, tags, query)
-
-    if staleness:
-        query = find_hosts_by_staleness(staleness, query)
-
-    try:
-        order_by = _params_to_order_by(order_by, order_how)
-    except ValueError as e:
-        flask.abort(400, str(e))
-    else:
-        query = query.order_by(*order_by)
-
-    query_results = query.paginate(page, per_page, True)
-    logger.debug("Found hosts: %s", query_results.items)
-
-    return _build_paginated_host_list_response(query_results.total, page, per_page, query_results.items)
-
-
-def _order_how(column, order_how):
-    if order_how == "ASC":
-        return column.asc()
-    elif order_how == "DESC":
-        return column.desc()
-    else:
-        raise ValueError('Unsupported ordering direction, use "ASC" or "DESC".')
-
-
-def _params_to_order_by(order_by=None, order_how=None):
-    modified_on_ordering = (Host.modified_on.desc(),)
-    ordering = ()
-
-    if order_by == "updated":
-        if order_how:
-            modified_on_ordering = (_order_how(Host.modified_on, order_how),)
-    elif order_by == "display_name":
-        if order_how:
-            ordering = (_order_how(Host.display_name, order_how),)
-        else:
-            ordering = (Host.display_name.asc(),)
-    elif order_by:
-        raise ValueError('Unsupported ordering column, use "updated" or "display_name".')
-    elif order_how:
-        raise ValueError(
-            "Providing ordering direction without a column is not supported. "
-            "Provide order_by={updated,display_name}."
+    total = 0
+    host_list = ()
+
+    config = inventory_config()
+    if config.bulk_query_source == BulkQuerySource.db:
+        try:
+            host_list, total = get_host_list_db(
+                display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
+            )
+        except ValueError as e:
+            flask.abort(400, str(e))
+    elif config.bulk_query_source == BulkQuerySource.xjoin:
+        host_list, total = get_host_list_xjoin(
+            display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
```

Comment:
given that the function signatures are identical there's no need to duplicate function invocation

```suggestion
    get_host_list = get_host_list_db if config.bulk_query_source == BulkQuerySource.db else get_host_list_xjoin

    try:
        host_list, total = get_host_list(
            display_name, fqdn, hostname_or_id, insights_id, tags, page, per_page, order_by, order_how, staleness
        )
    except ValueError as e:
        flask.abort(400, str(e))
```

Reply:
You mean like, inline the contents of that function? I slightly disagree here. There is quite a lot of logic in both function and having them extracted improves testability. Also there is a nice symmetry between the _db_ and _xjoin_ variants.

I‘d rather leave it as it is for now, but I agree that there is a room for improvement. Let’s rather leave that for an explicit refactoring. There is more of the code shuffling and unspaghetizing that would this code utilize.

Reply:
@Glutexo no, inlining is not at all what I meant. Take a look at the code suggestion.

Reply:
Oh, I see. The signature is not exactly same. Unfortunatly they differ in what exceptions they can trigger, which in my eyes is a part of the signature. I was thinking about refactoring that to introduce the exact change you suggest. I decided to keep that for some further refactoring cycle though.

If you insist on the refactoring, I can include it in this pull request. I wouldn’t want to do it the simplest way though: having two _except_ blocks for the same call would result in one of them always not being reachable. That seems like an antipattern to me.

Reply:
How do they differ in exceptions? From what I can see they both can raise `ValueError`. Not handling it for `get_host_list_xjoin` seems like an oversight not me.

Reply:
😯 You’re right. I’d swear I had a reason for that. This is worth adding a regression test.